sympy/matrices/matrices.py

```diff
@@ -3313,6 +3383,187 @@ def cleanup():
+        >>> J
+        Matrix([
+        [2, 1, 0, 0],
+        [0, 2, 0, 0],
+        [0, 0, 2, 1],
+        [0, 0, 0, 2]])
+
+        See Also
+        ========
+
+        jordan_block
+        """
+        if not self.is_square:
+            raise NonSquareMatrixError("Only square matrices have Jordan forms")
+
+        chop = kwargs.pop('chop', False)
+        mat = self
+        has_floats = any(v.has(Float) for v in self)
+
+        def restore_floats(*args):
+            """If `has_floats` is `True`, cast all `args` as
+            matrices of floats."""
+            if has_floats:
+                args = [m.evalf(chop=chop) for m in args]
+            if len(args) == 1:
+                return args[0]
+            return args
+
+        # cache calculations for some speedup
+        mat_cache = {}
+        def eig_mat(val, pow):
+            """Cache computations of (self - val*I)**pow for quick
+            retrieval"""
+            if (val, pow) in mat_cache:
+                return mat_cache[(val, pow)]
+            if (val, pow - 1) in mat_cache:
+                mat_cache[(val, pow)] = mat_cache[(val, pow - 1)] * mat_cache[(val, 1)]
+            else:
+                mat_cache[(val, pow)] = (mat - val*self.eye(self.rows))**pow
+            return mat_cache[(val, pow)]
+
+        # helper functions
+        def nullity_chain(val):
+            """Calculate the sequence  [0, nullity(E), nullity(E**2), ...]
+            until it is constant where `E = self - val*I`"""
+            # mat.rank() is faster than computing the null space,
+            # so use the rank-nullity theorem
+            cols = self.cols
+            ret = [0]
+            nullity = cols - eig_mat(val, 1).rank()
+            i = 2
+            while nullity != ret[-1]:
+                ret.append(nullity)
+                nullity = cols - eig_mat(val, i).rank()
+                i += 1
+            return ret
+
+        def blocks_from_nullity_chain(d):
+            """Return a list of the size of each Jordan block.
+            If d_n is the nullity of E**n, then the number
+            of Jordan blocks of size n is
+
+                2*d_n - d_(n-1) - d_(n+1)"""
+            # d[0] is always the number of columns, so skip past it
+            mid = [2*d[n] - d[n - 1] - d[n + 1] for n in range(1, len(d) - 1)]
+            # d is assumed to plateau with "d[ len(d) ] == d[-1]", so
+            # 2*d_n - d_(n-1) - d_(n+1) == d_n - d_(n-1)
+            end = [d[-1] - d[-2]] if len(d) > 1 else [d[0]]
+            return mid + end
+
+        def pick_vec(small_basis, big_basis):
+            """Picks a vector from big_basis that isn't in
+            the subspace spanned by small_basis"""
+            if len(small_basis) == 0:
+                return big_basis[0]
+            for v in big_basis:
+                _, pivots = self.hstack(*(small_basis + [v])).echelon_form(with_pivots=True)
+                if pivots[-1] == len(small_basis):
```

Comment:
This is a nice method. It should work over any (algebraically closed) field independently of its characteristic unlike the old method. (Assuming that the eigenvalues can be found.)